draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+        # snip, we just discussed this
+      end
+    end
+  end
+end
+```
+
+you can see that we actually call `env` on our `Env::Bootstrap` object:
+
+```Ruby
+  def env
+    subjects = matched_subjects
+    Env.new(
+      actor_env:        Actor::Env.new(Thread),
+      config:           config,
+      integration:      @integration,
+      matchable_scopes: matchable_scopes,
+      mutations:        subjects.flat_map(&:mutations),
+      parser:           parser,
+      selector:         Selector::Expression.new(@integration),
+      subjects:         subjects
+    )
+  end
+```
+
+This now brings us to the topic of "subjects".
+
+### Subjects
+
+Let's go through the `env` method step by step:
+
+```Ruby
+  def env
+    subjects = matched_subjects
+    # snip
+  end
+```
+
+First we'll get all "matched subjects", which means all the subjects you want to mutation tests. I won't go into detail how the `matched_subjects` method works because that would probably deserve another blog post so let's just assume it returns your modules and classes you would like to mutation test. 
+
+Let's look at an example - Remember the `Greeter` example from above?
+
+```
+bundle exec mutant --include lib/\
+                   --require hello_world\
+                   --use rspec  "Greeter*"
+```
+
+So given we fire up `Mutant` like that, how does `subjects` look like?
+`matched_subjects` returns an array of `Enumerable<Subject>`, let's check out how this looks in our console:
+
+```
+[
+# The `subjects` array has 2 entries:
+# First subject:
+
+#<Mutant::Subject::Method::Instance 
+  context=
+    #<Mutant::Context 
+      scope=Greeter
+      source_path=#<Pathname:.../hello_world/lib/hello_world/greeter.rb>>
+  node=s(:def, :initialize,
+    s(:args,
+      s(:arg, :phrase)),
+        s(:ivasgn, :@phrase,
+          s(:lvar, :phrase)))
+
+>, # end of first subject
+
+# Second subject:
+
+#<Mutant::Subject::Method::Instance
+  context=
+    #<Mutant::Context
+      scope=Greeter
+      source_path=#<Pathname:.../hello_world/lib/hello_world/greeter.rb>>
+  node= s(:def, :say_hello,
+          s(:args,
+            s(:arg, :name)),
+          s(:if,
+            s(:ivar, :@enabled),
+            s(:dstr,          
+              # ...snip
+    > # end of second subject
+]
+```
+
+Just by reading through it you should see be able to get a feeling for what's happening, but let's go through it step by step:
+
+```
+Mutant::Subject::Method::Instance
+```
+
+The `Subject` class. As you can see it has 2 attributes:
+
+1.) A [context](https://github.com/mbj/mutant/blob/master/lib/mutant/context.rb)
+
+```
+  context=
+    #<Mutant::Context 
+      scope=Greeter
+      source_path=#<Pathname:.../hello_world/lib/hello_world/greeter.rb>>
+```
+
+which contains:
+  * a [scope](https://github.com/mbj/mutant/blob/master/lib/mutant/scope.rb) called `Greeter`. This should look familiar to you now, doesn't it? That's the class from above we're actually testing.
```

Comment:
Actually its the namespace, not the class (as it could also be a module), I usually use "namespace" where what I'm talking about is either a class or a module.